### DISCUSSION TITLE
How to use resources and assets with external API call

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#8845] [Q&A] [answered]
Adapted from Slack: https://dagster.slack.com/archives/C01U954MEER/p1657555490945469?thread_ts=1657440754.738839&cid=C01U954MEER.

My project consists of pulling Spotify tracks via their api, extracting the artists from the songs and comparing them to concert venues to find potential gigs, scheduled weekly.

With the "prod" version of this, I'd want to run multiple api calls to Spotify with pagination. With the "local" version for testing, I'd ideally just load in a csv of the songs.

I want to do some data extraction to get all the unique artists to compare to venue gigs.

My understanding is that because Spotify would be a API call, then it should be a Resource? Which is then utilised within an Asset?

**Comment by sryza** [comment, marked as answer]
Here's an example of what this could look like, with the resource implementations left as an exercise to the reader. To run it with the mock spotify implementation, you would set the environment variable DAGSTER_ENVIRONMENT=test.

```
import os
from dagster import Definitions, asset, resource, load_assets_from_current_module
from pandas import DataFrame


class Spotify:
    def fetch_tracks(self) -> DataFrame:
        ...


class MockSpotify:
    def fetch_tracks(self) -> DataFrame:
        ...


@resource
def spotify_resource():
    return Spotify()


@resource
def mock_spotify_resource():
    return MockSpotify()


@asset(required_resource_keys={"spotify"})
def spotify_tracks(context) -> DataFrame:
    return context.resources.spotify.fetch_tracks()


@asset
def unique_artists(spotify_tracks: DataFrame):
    return spotify_tracks["artist"].unique().tolist()


@asset
def spotify_group_artists_track_count(spotify_tracks: DataFrame) -> DataFrame:
    spotify_group_artists_track_count_df = (
        spotify_tracks.groupby(["artist"])["track_id"]
        .nunique()
        .reset_index(name="count")
        .sort_values(by="count", ascending=False)
    )
    return spotify_group_artists_track_count_df


environment = os.environ.get("DAGSTER_ENVIRONMENT", "prod")
prod_resources = {"spotify": spotify_resource}
test_resources = {"spotify": mock_spotify_resource}
resource_defs = {"prod": prod_resources, "test": test_resources}[environment]

defs = Definitions(assets=load_assets_from_current_module(), resources=resource_defs)
```

**Reply by ghost** [reply]
Very usefull to understand how to use resources. Thank you !
